### PR TITLE
Implement Blanx Training

### PR DESCRIPTION
**Pull request by maciek-pioro**
# Description
todo

# Checklist
   - [ ] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [ ] Are all variables set with logical defaults that are backward compatible?
   - [ ] I attached link to neptune if it was necessary
   - [ ] Have you successfully executed the linter and tests?
   - [ ] Are all functions concise and don't require splitting?
   - [ ] Is there documentation for arguments and complex logic?
   - [ ] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [ ] Are all functions placed in appropriate files?
   - [ ] Is the PR name meaningful and descriptive?
   - [ ] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [ ] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [ ] If there is a need for a second reviewer I requested additional review

### CODE REVIEW

File: lizrd/text/packers.py [hunk 1]
```diff
@@ -206,3 +206,100 @@ def get_sample(self) -> LLMExample:
...
+            #     )
+
+            # blank_insertion_point + i - 1 + 1 + self.n_blanks - i + len(tokens) - blank_insertion_point
+            # len(tokens) + blanks
+            buffer_input.extend(input_tokens)
+            buffer_output.extend(output_tokens)
+            blank_mask_buffer.extend(blank_mask)
+
+            document_lengths.append(len(tokens))
+            if (sum(document_lengths) - max(document_lengths)) > self.sequence_length:
```
Comment (sebastianjaszczur) [inline lizrd/text/packers.py:282]: Jeśli byśmy w jakiś sposób liczyli sumę dokumentów nie licząc blanków, to mielibyśmy, chyba, te same dokumenty co baseline? do dyskusji

File: lizrd/text/packers.py [hunk 2]
```diff
@@ -206,3 +206,100 @@ def get_sample(self) -> LLMExample:
...
+        assert len(buffer_input) == len(buffer_output) == len(blank_mask_buffer)
+
+        illegal_blanks_position = True
+        while illegal_blanks_position:
+            sample_start = self.py_rng.randint(0, len(buffer_input) - 1)
+            sample_end = sample_start + self.sequence_length
+            blanks_at_beginning = blank_mask_buffer[sample_start] == 1
+            # fix bug
+            blanks_at_end = sum(
+                take_circular(blank_mask_buffer, sample_end - self.n_blanks, sample_end)
```
Comment (sebastianjaszczur) [inline lizrd/text/packers.py]: what? czemu nie sprawdzać po prostu czy index `blank_mask_buffer[sample_end] == 1`? Czy tam `blank_mask_buffer[sample_end] == 1`

Comment (sebastianjaszczur) [reply lizrd/text/packers.py]: W wersji hardcorowej, hack, ale chyba nie polecałbym (bo nie będzie działać przy randomizowanej liczbie blanków) - sprawdzamy czy blanków nie ma na początku, i sprawdzamy czy `sum(blank_mask_buffer[sample_start:sample_end]) %n_blanks == 0`

File: lizrd/text/tokenizers.py [hunk 3]
```diff
@@ -37,12 +38,16 @@ def text_to_ids(self, text: str) -> List[int]:
 
 
 class GPTTokenizer(AbstractTokenizer):
-    VOCAB_SIZE = 50257
```
Comment (sebastianjaszczur) [inline lizrd/text/tokenizers.py:40]: waaaait, czy my więc zmieniamy wszystkie modele? to mi się nie podoba jakoś za bardzo

Comment (sebastianjaszczur) [reply lizrd/text/tokenizers.py:40]: czy nie ma w tym vocabie juz jakiś specjalnych tokenów które można użyć, maybe?

Comment (sebastianjaszczur) [reply lizrd/text/tokenizers.py:40]: no, ale jeśli trzeba to spoko. Trochę się jednak boję o kompatybilność wsteczną...

File: research/conditional/utils/conditional_trainer.py [hunk 4]
```diff
@@ -135,13 +138,13 @@ def train(self, n_steps: int):
...
                 print(f"Step {step}")
 
-            if (
-                self.model_type == "gpt"
-                and self.decoding_logging_steps > 0
-                and step % self.decoding_logging_steps == 0
-                and self.is_process_logging
-            ):
-                self._decode_samples(step)
+            # if (
```
Comment (sebastianjaszczur) [inline research/conditional/utils/conditional_trainer.py:141]: Yo, the desired solution I believe is to set decoding_logging_steps to 0 to disable decoding.

File: research/conditional/utils/model_utils.py [hunk 5]
```diff
@@ -129,6 +134,86 @@ def calculate_llm_loss(
...
+            # print(blank_start.shape)
+            nth_blank_mask = F.pad(blank_start[:, 1:], (0, 1), "constant", 0)
+            # print(nth_blank_mask.shape)
+            for i in range(n_blanks + 1):
+                if i == 0:
+                    assert not input_tokens[nth_blank_mask == 1].eq(blank_id).any()
+                else:
+                    assert input_tokens[nth_blank_mask == 1].eq(blank_id).all()
+                nth_blank_loss = (nth_blank_mask.reshape(-1) * mask_loss).mean()
+                blanks_losses[f"blank_{i}_loss"] = nth_blank_loss
```
Comment (sebastianjaszczur) [inline research/conditional/utils/model_utils.py:193]: adding a diff losses would be great. In particular, for every (positive) blank i, compute diff(i, 0) and diff(i, i-1).

Comment (sebastianjaszczur) [reply research/conditional/utils/model_utils.py:193]: not necessary for MVP, maybe, but it'll make it all easier to analyze

File: research/blanks/model.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        self.blank_weight = nn.Parameter(torch.tensor(initial_blank_weight))
+        self.use_straight_through = use_straight_through
+
+    def forward(self, encoder_output: torch.Tensor, model_input: torch.Tensor):
+        is_blank = model_input.eq(self.blank_token_id)
+        is_first_blank = get_first_blanks_in_series(is_blank)
+        is_preblank = shift_left(is_first_blank)
+        preblank_encoder_output = encoder_output * is_preblank.unsqueeze(-1)
+        if self.learnable_weights:
+            is_not_blank = ~is_blank
```
Comment (maciek-pioro) [inline research/blanks/model.py:126]: Dać asserta

**Comment by sebastianjaszczur** [comment]
When do we want to merge 🅱️ ?

File: lizrd/support/misc.py [hunk 7]
```diff
@@ -3,6 +3,8 @@
 import random
 import string
 from typing import Optional, List, Tuple, Set
+import numpy as np
+import torch
```
Comment (maciek-pioro) [inline lizrd/support/misc.py:7]: Zobaczyć czy to się nie wywala bez venva

File: lizrd/text/tokenizers.py [hunk 8]
```diff
@@ -8,6 +8,7 @@ class AbstractTokenizer(ABC):
     sequence_separator_id: Optional[int]
     mask_id: Optional[int]
     eot_id: Optional[int]
+    blank_id: Optional[int]
```
Comment (maciek-pioro) [inline lizrd/text/tokenizers.py:11]: Wrzucić do trello jako dług techniczny (że powinna być bardziej skomplikowana hierarchia tokenizerów).

File: research/blanks/argparse.py
Comment (maciek-pioro) [inline research/blanks/argparse.py:1]: Spróbować tutaj wywalić niepotrzebne

File: research/blanks/configs/example.yaml
Comment (maciek-pioro) [inline research/blanks/configs/example.yaml:1]: Zostawić tylko jeśli ma się odpalać w testach

File: research/blanks/configs/less_embeddings.yaml
Comment (maciek-pioro) [inline research/blanks/configs/less_embeddings.yaml:1]: jw

File: research/blanks/data.py
Comment (maciek-pioro) [inline research/blanks/data.py:1]: Rename to datasets

File: research/blanks/loss.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+)
+
+
+def make_loss_function(loss_checkpoint_chungs: int, n_blanks: int = 0):
+    if loss_checkpoint_chungs == 0:
+        return partial(calculate_llm_loss, n_blanks=n_blanks)
+    else:
+        if n_blanks > 0:
+            raise NotImplementedError(
+                "Loss checkpointing not supported for blanks (yet)"
```
Comment (maciek-pioro) [inline research/blanks/loss.py:22]: Loss chungization not checkpointing

**Review by crewtool — APPROVED** [review]
Usiedliśmy z Maćkiem i mi tłumaczył kod, tak jak go rozumiem brzmi dobrze ;)

File: research/blanks/argparse.py
Comment (maciek-pioro) [reply research/blanks/argparse.py:1]: Dodane do trello

File: research/blanks/model.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+    blanks_learnable_weights: bool = False,
+    blank_initial_weight: float = 1.0,
+    blanks_straight_through: bool = False,
+):
+    if model_fragmentation is None or device == torch.device("cpu"):
+        first_gpu = device
+        last_gpu = device
+    else:
+        first_gpu = torch.device("cuda:0")
+        last_gpu = torch.device(f"cuda:{len(model_fragmentation)}")
```
[the remark this answers is not in the document]
Comment (maciek-pioro) [reply research/blanks/model.py:36]: Not related to this PR

File: research/blanks/model.py [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+        self.blank_token_id = blank_token_id
+        self.n_blanks = n_blanks
+
+        self.learnable_weights = learnable_weights
+        self.preblank_weight = nn.Parameter(torch.tensor(1.0))
+        self.blank_weight = nn.Parameter(torch.tensor(initial_blank_weight))
+        self.use_straight_through = use_straight_through
+
+    ...
+
```
[the remark this answers is not in the document]
Comment (maciek-pioro) [reply research/blanks/model.py:171]: Will be used in another PR

File: research/blanks/model.py [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+            n_blanks=n_blanks,
+            learnable_weights=blanks_learnable_weights,
+            initial_blank_weight=blank_initial_weight,
+            use_straight_through=blanks_straight_through,
+        ).to(last_gpu)
+    else:
+        head = llm.PredictionHead(dm, vocab_size).to(last_gpu)
+
+    if n_blanks > 0:
+        model = research.blanks.model.BlankLLM(embedding_layer, encoder_tower, head)
```
[the remark this answers is not in the document]
Comment (maciek-pioro) [reply research/blanks/model.py:77]: thx
